tests/frontend/test_io.py

```diff
@@ -0,0 +1,543 @@
+        sf.save(filename, prog)
+
+        with open(filename, "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+    def test_save_file_object(self, prog, tmpdir):
+        """Test writing a program to a file object"""
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            sf.save(f, prog)
+
+        with open(filename, "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+    def test_save_append_extension(self, prog, tmpdir):
+        """Test appending the extension if not present"""
+        filename = str(tmpdir.join("test.txt"))
+        sf.save(filename, prog)
+
+        with open(filename+'.xbb', "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+
+class TestLoad:
+    """Test sf.load functionality"""
+
+    def test_invalid_file(self, prog, tmpdir):
+        """Test exception is raised if file is invalid"""
+        with pytest.raises(ValueError, match="must be a string, path"):
+            sf.load(1)
+
+    def test_load_filename_path_object(self, prog, tmpdir):
+        """Test loading a program using a path object"""
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            f.write(test_prog_not_compiled)
+
+        res = sf.load(filename)
+
+        # check loaded program is the same as expected
+        for cmd1, cmd2 in zip(res.circuit, prog.circuit):
+            assert cmd1.op.__class__.__name__ == cmd2.op.__class__.__name__
+            assert cmd1.reg[0].ind == cmd2.reg[0].ind
+            if cmd1.op.p:
+                assert np.all(cmd1.op.p[0].x == cmd2.op.p[0].x)
+
+    def test_load_filename_string(self, prog, tmpdir):
+        """Test loading a program using a string filename"""
+        filename = str(tmpdir.join("test.xbb"))
+
+        with open(filename, "w") as f:
+            f.write(test_prog_not_compiled)
+
+        res = sf.load(filename)
+
+        # check loaded program is the same as expected
+        for cmd1, cmd2 in zip(res.circuit, prog.circuit):
+            assert cmd1.op.__class__.__name__ == cmd2.op.__class__.__name__
+            assert cmd1.reg[0].ind == cmd2.reg[0].ind
+            if cmd1.op.p:
+                assert np.all(cmd1.op.p[0].x == cmd2.op.p[0].x)
+
+    def test_load_file_object(self, prog, tmpdir):
+        """Test loading a program via a file object"""
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            sf.save(f, prog)
+
+        with open(filename, "r") as f:
+            res = sf.load(f)
+
+        # check loaded program is the same as expected
+        for cmd1, cmd2 in zip(res.circuit, prog.circuit):
+            assert cmd1.op.__class__.__name__ == cmd2.op.__class__.__name__
+            assert cmd1.reg[0].ind == cmd2.reg[0].ind
+            if cmd1.op.p:
+                assert np.all(cmd1.op.p[0].x == cmd2.op.p[0].x)
```

Comment:
probably worth writing a separate function for this to avoid code duplication

Reply:
done